### PR TITLE
feat : WooCommerce Integration

### PR DESCRIPTION
**Pull request by aaronmenezes**
- [x] Basic migration (doctypes, utils)
- [x] Use ecommerce item instead of custom fields
- [x] Write patch to migrate existing users
- [ ] Remove integration from core

### CODE REVIEW

File: ecommerce_integrations/woocommerce/woocommerce_connection.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+	default_warehouse = _("Stores - {0}", sys_lang).format(company_abbr)
+	if not frappe.db.exists("Warehouse", default_warehouse) and not woocommerce_settings.warehouse:
+		frappe.throw(_("Please set Warehouse in Woocommerce Setting"))
+
+	for item in order.get("line_items"):
+		woocomm_item_id = item.get("product_id")
+		found_item = frappe.get_doc("Item", {"woocommerce_id": cstr(woocomm_item_id)})
+
+		ordered_items_tax = item.get("total_tax")
```
Comment (ankush) [inline ecommerce_integrations/woocommerce/woocommerce_connection.py:231]: This line should be moved nearer to where it's actually being used.

File: ecommerce_integrations/woocommerce/woocommerce_connection.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+		)
+
+	# shipping_details = order.get("shipping_lines") # used for detailed order
+
+	add_tax_details(
+		new_sales_order, order.get("shipping_tax"), "Shipping Tax", woocommerce_settings.f_n_f_account
+	)
+	add_tax_details(
+		new_sales_order,
+		order.get("shipping_total"),
```
Comment (ankush) [inline ecommerce_integrations/woocommerce/woocommerce_connection.py:258]: Logically shipping costs should be added before shipping tax.

**Review by ankush — CHANGES_REQUESTED** [review]

File: ecommerce_integrations/woocommerce/woocommerce_connection.py (hunk 2 above)
Comment (ankush) [reply ecommerce_integrations/woocommerce/woocommerce_connection.py:258]: Also IMO we should review tax computation thoroughly. There were several major issues with the current Shopify integration and I think this implementation of woocommerce taxes feels way to simple to be correct.

File: ecommerce_integrations/woocommerce/doctype/woocommerce_setting/woocommerce_setting.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	def validate(self):
+		self.validate_settings()
+		self.create_delete_custom_fields()
+		self.create_webhook_url()
+		if self.enable_sync:
+			create_ecommerce_items()
+
+	def create_delete_custom_fields(self):
+		if self.enable_sync:
+			for doctype in ["Customer", "Sales Order", "Item", "Address"]:
```
Comment (ankush) [inline ecommerce_integrations/woocommerce/doctype/woocommerce_setting/woocommerce_setting.py:25]:
```suggestion
			for doctype in ["Customer", "Sales Order", "Address"]:
```

If e-commerce item is being used then Item should not need a custom field now.

File: ecommerce_integrations/woocommerce/doctype/woocommerce_setting/woocommerce_setting.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
+# Copyright (c) 2021, Frappe and contributors
+# For license information, please see license.txt
+
+import frappe
+from frappe import _
+from frappe.custom.doctype.custom_field.custom_field import create_custom_field
+from frappe.model.document import Document
+from frappe.utils.nestedset import get_root_of
+from six.moves.urllib.parse import urlparse
```
[the remark this answers is not in the document]
Comment (ankush) [reply ecommerce_integrations/woocommerce/doctype/woocommerce_setting/woocommerce_setting.py:9]: removed now: https://github.com/frappe/frappe/pull/15763

File: ecommerce_integrations/woocommerce/woocommerce_connection.py [hunk 5]
```diff
@@ -0,0 +1,273 @@
+import base64
+import hashlib
+import hmac
+import json
+
+import frappe
+from frappe import _
+from frappe.utils import cstr
```
Comment (ankush) [inline ecommerce_integrations/woocommerce/woocommerce_connection.py:8]:
```suggestion
from frappe.utils import cstr, flt
```

File: ecommerce_integrations/woocommerce/woocommerce_connection.py [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+	new_sales_order.insert(ignore_permissions=True)
+	new_sales_order.submit()
+
+
+def set_items_in_sales_order(new_sales_order, woocommerce_settings, order, sys_lang):
+	company_abbr = frappe.db.get_value("Company", woocommerce_settings.company, "abbr")
+
+	default_warehouse = _("Stores - {0}", sys_lang).format(company_abbr)
+	if not frappe.db.exists("Warehouse", default_warehouse) and not woocommerce_settings.warehouse:
+		frappe.throw(_("Please set Warehouse in Woocommerce Setting"))
```
[the remark this answers is not in the document]
Comment (aaronmenezes) [reply ecommerce_integrations/woocommerce/woocommerce_connection.py]: Will remove this check and make Warehouse selection mandatory in woocommerce setting.

**Pull request closed without merging**